### DISCUSSION TITLE
[Feature Request] Add support for Cloud Functions v2

### DISCUSSION THREAD

**Discussion by atanaskanchev** [invertase/react-native-firebase#6543] [Ideas] [answered]
Currently trying to call a v2 callable function with `functions().httpsCallable('my-v2-fn-name')(payload)` I get an error `INTERNAL` coming from `@react-native-firebase/functions`

**Comment by mikehardy** [comment]
Hi there! Did you see anything in the native logs? Just curious. Also, what versions are you on, etc etc.

**Reply by atanaskanchev** [reply]
Hi @mikehardy I am using the following packages:

>     "@react-native-firebase/app": "^15.5.0",
>     "@react-native-firebase/app-check": "^15.5.0",
>     "@react-native-firebase/auth": "^15.5.0",
>     "@react-native-firebase/crashlytics": "^15.5.0",
>     "@react-native-firebase/firestore": "^15.5.0",
>     "@react-native-firebase/functions": "^15.5.0",
>     "@react-native-firebase/perf": "^15.5.0",
>     "@react-native-firebase/storage": "^15.5.0",

>     "firebase": "9.10.0",
>     "firebase-admin": "^11.0.1",
>     "firebase-functions": "^3.23.0",

Unfortunately there isn't any useful information in the stack trace, it's just internal error.

In the firebase docs https://firebase.google.com/docs/functions/beta/get-started the v2 callable functions must be called via `import { getFunctions, httpsCallableFromURL } from 'firebase/functions';`, so I suppose it won't work as a standard v1 callable function

**Reply by mikehardy** [reply]
I see, that's a firebase v9 API mostly but have a need for getcallablefromurl as well, and we need to support v2 functions so I will see what I can do

**Reply by atanaskanchev** [reply]
The strange thing is `functions().httpsCallable('my-v2-fn-name')(payload)` works fine with a v2 callable function running in the local emulator, but it fails with a deployed one.

**Reply by mikehardy** [reply]
that is really great to know! most of our work is with the emulator. It would be good to have a real/complete code snippet of exactly how you are calling this, to inform any future testing. You can post it in here with triple-backticks and a language type marker like

```javascript
  // this should be some code
  let x = undefined;
```

**Comment by atanaskanchev** [comment]
The function code:

```ts
import { FieldValue } from 'firebase-admin/firestore'
import { logger } from 'firebase-functions/v2'
import { CallableRequest, HttpsError, onCall } from 'firebase-functions/v2/https'
import { v4 as uuidv4 } from 'uuid'

const fn = onCall(async (request: CallableRequest) => {
  const { data, auth } = request

  // ? app check
  // if (!context.app) {
  //   throw new https.HttpsError('failed-precondition', 'unauthorized request')
  // }

  // ? auth check
  if (!auth) {
    throw new HttpsError('unauthenticated', 'unauthenticated request')
  }

  const { uid } = auth
  const { customer_id } = auth.token

  const id = uuidv4()
  const colRef = 'target_collection'
  const docRef = colRef.doc(id)

  await docRef
    .create({
      ...data,

      meta: {
        customer_id,
        created_at: FieldValue.serverTimestamp(),
        created_by: uid,
        id,
        audience: [...new Set<string>([uid])]
      }
    })
    .catch((error) => {
      const error_id = uuidv4()
      logger.error(new Error('v2 fn error'), error, { error_id })
      throw new HttpsError('internal', { error_id })
    })

  return { id }
})
```

and call it in react native as `functions().httpsCallable('the-v2-fn-name')(payload)`

**Reply by mikehardy** [reply]
I was actually hoping more for the calling code :laughing: - focused on the client more - just to make sure I know exactly how to reproduce the behavior you saw as works on emulator / does not work on deployed function since that seems subtle (and possibly like an emulator bug? That is, the emulator is not bug-compatible with cloud - which makes me curious what version of firebase-tools you've got on package-lock.json or yarn.lock

Sorry for all the picky questions but the details sometimes make all the difference, I appreciate the help

**Reply by atanaskanchev** [reply]
`firebase-tools@11.9.0`

The emulator setup is 
```ts
import auth from '@react-native-firebase/auth'
import firestore from '@react-native-firebase/firestore'
import functions from '@react-native-firebase/functions'
import storage from '@react-native-firebase/storage'

 if (__DEV__ && isLoadingComplete) {
    functions().useEmulator('localhost', 5001)
    firestore().useEmulator('localhost', 8080)
    storage().useEmulator('localhost', 9199)
    auth().useEmulator('http://localhost:9099')
  }
```

The emulator initialises as:
`  functions: Loaded functions definitions from source: v2.my.function.name.
✔  functions[us-central1-v2-my-function-name]: http function initialized (http://localhost:5001/my-project1/v2-my-function-name).
`

And how I call it
```ts
import functions, { FirebaseFunctionsTypes } from '@react-native-firebase/functions'
import { useMutation, UseMutationResult } from '@tanstack/react-query'

export const useCreateApplianceFn = (): UseMutationResult<
  FirebaseFunctionsTypes.HttpsCallableResult,
  Error,
  unknown,
  { id: string }
> =>
  useMutation(['mutation-key'], (payload) => {
    return functions().httpsCallable('v2-my-function-name')(payload)
  })

export default useCreateApplianceFn

```

**Reply by evelant** [reply]
@mikehardy FYI I found that v2 functions work fine _on the emulators_ even on lower versions of the firebase libs. I was on RNFirebase v14 (to avoid the use_frameworks! issues for now) and wrote v2 functions using the emulator without issue. I got surprised when I deployed and discovered that they don't work remotely!

**Comment by mikehardy** [comment, marked as answer]
#6626 implements `getCallableFromUrl(url)` and I'll merge it and release as soon as CI goes green
